providers/azure_test.go

```diff
@@ -122,98 +157,154 @@ func testAzureBackend(payload string) *httptest.Server {
+			ExpectedError:           fmt.Errorf("unable to get email address, error: %v", errors.New("type assertion to string failed")),
+		},
+		{
+			Description:   "should not query profile api when email is already set in session",
+			Email:         "user@windows.net",
+			ExpectedEmail: "user@windows.net",
+		},
+	}
 
-	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
-	assert.Equal(t, nil, err)
-	assert.Equal(t, "", email)
+	for _, testCase := range testCases {
+		t.Run(testCase.Description, func(t *testing.T) {
+			var (
+				b    *httptest.Server
+				host string
+			)
+			if testCase.PayloadFromAzureBackend != "" {
+				b = testAzureBackend(testCase.PayloadFromAzureBackend)
+				defer b.Close()
+
+				bURL, _ := url.Parse(b.URL)
+				host = bURL.Host
+			}
+			p := testAzureProvider(host)
+			session := CreateAuthorizedSession()
+			session.Email = testCase.Email
+			err := p.EnrichSession(context.Background(), session)
+			assert.Equal(t, testCase.ExpectedError, err)
+			assert.Equal(t, testCase.ExpectedEmail, session.Email)
+		})
+	}
 }
 
-func TestAzureProviderGetEmailAddressIncorrectOtherMails(t *testing.T) {
-	b := testAzureBackend(`{ "mail": null, "otherMails": "", "userPrincipalName": null }`)
-	defer b.Close()
-
-	bURL, _ := url.Parse(b.URL)
-	p := testAzureProvider(bURL.Host)
-
-	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
-	assert.Equal(t, "type assertion to string failed", err.Error())
-	assert.Equal(t, "", email)
-}
+func TestAzureProviderRedeem(t *testing.T) {
+	testCases := []struct {
+		Name                 string
+		RefreshToken         string
+		ExpiresOn            string
+		EmailFromIDToken     string
+		EmailFromAccessToken string
+		IsIDTokenMalformed   bool
+	}{
+		{
+			Name:             "with id_token returned",
+			EmailFromIDToken: "foo1@example.com",
+			RefreshToken:     "some_refresh_token",
+			ExpiresOn:        "2006-01-02T22:04:05Z",
+		},
+		{
+			Name:                 "without id_token returned, fallback to access token",
+			EmailFromAccessToken: "foo2@example.com",
+			RefreshToken:         "some_refresh_token",
+			ExpiresOn:            "2006-01-02T22:04:05Z",
+		},
+		{
+			Name:                 "id_token malformed, fallback to access token",
+			EmailFromAccessToken: "foo3@example.com",
+			RefreshToken:         "some_refresh_token",
+			ExpiresOn:            "2006-01-02T22:04:05Z",
+			IsIDTokenMalformed:   true,
+		},
+	}
 
-func TestAzureProviderRedeemReturnsIdToken(t *testing.T) {
-	b := testAzureBackend(`{ "id_token": "testtoken1234", "expires_on": "1136239445", "refresh_token": "refresh1234" }`)
-	defer b.Close()
-	timestamp, err := time.Parse(time.RFC3339, "2006-01-02T22:04:05Z")
-	assert.Equal(t, nil, err)
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			idTokenString := ""
+			accessTokenString := ""
+			if testCase.EmailFromIDToken != "" {
```

Comment:
2 test cases I think need adding to get better coverage:

- Email in both a valid IDToken & Access Token
- Call to Redeem URL fails and results in an error

Reply:
sure